docs/doxygen/man/man3/GetAccounts.3
.TH "GetAccounts" 3 "Thu Aug 30 2018" "esc" \" -*- nroff -*-
.ad l
.nh
.SH NAME
GetAccounts \- 
.SH SYNOPSIS
.br
.PP
.PP
Inherits \fBBlockCommand\fP\&.
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBGetAccounts\fP (uint16_t src_bank, uint32_t src_user, uint32_t block, uint16_t dst_bank, uint32_t time)"
.br
.ti -1c
.RI "\fBGetAccounts\fP (const \fBGetAccounts\fP &obj)=delete"
.br
.RI "\fIDisabled copy contructor\&. \fP"
.ti -1c
.RI "\fBGetAccounts\fP & \fBoperator=\fP (const \fBGetAccounts\fP &)=delete"
.br
.RI "\fIDisabled copy assignment operator\&. \fP"
.ti -1c
.RI "virtual int \fBgetType\fP () override"
.br
.RI "\fIReturn TXSTYPE_NOD as type \&. \fP"
.ti -1c
.RI "virtual CommandType \fBgetCommandType\fP () override"
.br
.RI "\fIReturn eReadingOnly as command type \&. \fP"
.ti -1c
.RI "virtual unsigned char * \fBgetData\fP () override"
.br
.RI "\fIGet pointer to command data structure\&. \fP"
.ti -1c
.RI "virtual unsigned char * \fBgetResponse\fP () override"
.br
.RI "\fIGet pointer to response data\&. \fP"
.ti -1c
.RI "virtual void \fBsetData\fP (char *data) override"
.br
.RI "\fIPut data as a char list and put convert it to data structure\&. \fP"
.ti -1c
.RI "virtual void \fBsetResponse\fP (char *response) override"
.br
.RI "\fIApply data as a response struct\&. \fP"
.ti -1c
.RI "virtual int \fBgetDataSize\fP () override"
.br
.RI "\fIGet data struct size\&. Without signature\&. \fP"
.ti -1c
.RI "virtual int \fBgetResponseSize\fP () override"
.br
.RI "\fIGet response data struct size\&. \fP"
.ti -1c
.RI "virtual unsigned char * \fBgetSignature\fP () override"
.br
.RI "\fIGet pointer to signature data\&. \fP"
.ti -1c
.RI "virtual int \fBgetSignatureSize\fP () override"
.br
.RI "\fIGet signature size\&. \fP"
.ti -1c
.RI "virtual void \fBsign\fP (const uint8_t *hash, const uint8_t *sk, const uint8_t *pk) override"
.br
.RI "\fISign actual data plus hash using user private and public keys\&. \fP"
.ti -1c
.RI "virtual bool \fBcheckSignature\fP (const uint8_t *hash, const uint8_t *pk) override"
.br
.RI "\fICheck actual signature\&. \fP"
.ti -1c
.RI "virtual uint32_t \fBgetTime\fP () override"
.br
.RI "\fIGet time of command\&. \fP"
.ti -1c
.RI "virtual uint32_t \fBgetUserId\fP () override"
.br
.RI "\fIGet User ID\&. \fP"
.ti -1c
.RI "virtual uint32_t \fBgetBankId\fP () override"
.br
.RI "\fIGet Node ID\&. \fP"
.ti -1c
.RI "virtual int64_t \fBgetFee\fP () override"
.br
.RI "\fIGet command fee\&. \fP"
.ti -1c
.RI "virtual int64_t \fBgetDeduct\fP () override"
.br
.RI "\fIGet change in cash balance after command\&. \fP"
.ti -1c
.RI "virtual uint32_t \fBgetUserMessageId\fP () override"
.br
.RI "\fIGet message id\&. \fP"
.ti -1c
.RI "virtual bool \fBsend\fP (\fBINetworkClient\fP &netClient) override"
.br
.RI "\fISend data to the server\&. \fP"
.ti -1c
.RI "virtual void \fBsaveResponse\fP (\fBsettings\fP &sts) override"
.br
.RI "\fISave command response to settings object\&. \fP"
.ti -1c
.RI "virtual std::string \fBtoString\fP (bool pretty) override"
.br
.ti -1c
.RI "virtual void \fBtoJson\fP (boost::property_tree::ptree &ptree) override"
.br
.ti -1c
.RI "virtual void \fBtxnToJson\fP (boost::property_tree::ptree &ptree) override"
.br
.ti -1c
.RI "virtual std::string \fBusageHelperToString\fP () override"
.br
.ti -1c
.RI "uint32_t \fBgetDestBankId\fP ()"
.br
.RI "\fIGet destination bank id\&. \fP"
.ti -1c
.RI "uint32_t \fBgetBlockId\fP ()"
.br
.RI "\fIGet block id\&. \fP"
.ti -1c
.RI "\fBErrorCodes::Code\fP \fBprepareResponse\fP (uint32_t lastPath, uint32_t lastUsers)"
.br
.RI "\fIPrepare response buffer\&. \fP"
.in -1c
.SS "Additional Inherited Members"
.SH "Detailed Description"
.PP 
Definition at line 9 of file getaccounts\&.h\&.
.SH "Member Function Documentation"
.PP 
.SS "bool GetAccounts::checkSignature (const uint8_t * hash, const uint8_t * pk)\fC [override]\fP, \fC [virtual]\fP"

.PP
Check actual signature\&. 
.PP
\fBParameters:\fP
.RS 4
\fIhash\fP Previous hash operation\&. 
.br
\fIpk\fP Pointer to public key\&. 
.RE
.PP

.PP
Implements \fBICommand\fP\&.
.PP
Definition at line 64 of file getaccounts\&.cpp\&.
.SS "bool GetAccounts::send (\fBINetworkClient\fP & netClient)\fC [override]\fP, \fC [virtual]\fP"

.PP
Send data to the server\&. 
.PP
\fBParameters:\fP
.RS 4
\fInetClient\fP Netwrok client implementation of \fBINetworkClient\fP interface\&. 
.br
\fIpk\fP Pointer to public key\&. 
.RE
.PP

.PP
Implements \fBICommand\fP\&.
.PP
Definition at line 91 of file getaccounts\&.cpp\&.
.SS "void GetAccounts::sign (const uint8_t * hash, const uint8_t * sk, const uint8_t * pk)\fC [override]\fP, \fC [virtual]\fP"

.PP
Sign actual data plus hash using user private and public keys\&. 
.PP
\fBParameters:\fP
.RS 4
\fIhash\fP Previous hash operation\&. 
.br
\fIsk\fP Pointer to provate key\&. 
.br
\fIpk\fP Pointer to public key\&. 
.RE
.PP

.PP
Implements \fBICommand\fP\&.
.PP
Definition at line 60 of file getaccounts\&.cpp\&.

.SH "Author"
.PP 
Generated automatically by Doxygen for esc from the source code\&.


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

## [1.0.0] - 2018-08-24
### Changed
- Genesis file for the launch
- Update README

## [0.0.6] - 2018-08-10
### Added
- Profit sharing among TOP and VIP nodes

### Changed
- Genesis file for test launch

### Fixed
- Vote counting in get_block command
- Ability to send externally signed transactions

## [0.0.5] - 2018-08-01
### Added
- Support passing version by cmake command.
- Contributing guide
- "Dry run" test case

### Fixed
- Node stops after incorrect `send_many` request  - #134
- Fee checking
- Internal transactions numbers

## [0.0.4] - 2018-07-27
### Changed
- Update default parameters
- Creating a dev version with reference to the last tag

[Unreleased]: https://github.com/adshares/ads/compare/v1.0.0...HEAD

[0.0.5]: https://github.com/adshares/ads/compare/v0.0.6...v1.0.0
[0.0.5]: https://github.com/adshares/ads/compare/v0.0.5...v0.0.6
[0.0.5]: https://github.com/adshares/ads/compare/v0.0.4...v0.0.5
[0.0.4]: https://github.com/adshares/ads/compare/v0.0.3...v0.0.4


src/common/command/getaccounts.cpp
#include <fcntl.h>
#include <sys/stat.h>
#include "getaccounts.h"
#include "ed25519/ed25519.h"
#include "abstraction/interfaces.h"
#include "helper/json.h"
#include "helper/blocks.h"
#include "helper/blockfilereader.h"

GetAccounts::GetAccounts()
    : m_data{} {
}

GetAccounts::GetAccounts(uint16_t src_bank, uint32_t src_user, uint32_t block, uint16_t dst_bank, uint32_t time)
    : m_data(src_bank, src_user, time, block, dst_bank) {
    if (!dst_bank) {
        m_data.info.dst_node = src_bank;
    }
}

unsigned char* GetAccounts::getData() {
    return reinterpret_cast<unsigned char*>(&m_data.info);
}

unsigned char* GetAccounts::getResponse() {
    return m_responseBuffer.get();
}

void GetAccounts::setData(char* data) {
    m_data = *reinterpret_cast<decltype(m_data)*>(data);
}

void GetAccounts::setResponse(char* /*response*/) {
}

int GetAccounts::getDataSize() {
    return sizeof(m_data.info);
}

int GetAccounts::getResponseSize() {
    return m_responseBufferLength;
}

unsigned char* GetAccounts::getSignature() {
    return m_data.sign;
}

int GetAccounts::getSignatureSize() {
    return sizeof(m_data.sign);
}

int GetAccounts::getType() {
    return TXSTYPE_NOD;
}

CommandType GetAccounts::getCommandType() {
    return CommandType::eReadingOnly;
}

void GetAccounts::sign(const uint8_t* /*hash*/, const uint8_t* sk, const uint8_t* pk) {
    ed25519_sign(getData(), getDataSize(), sk, pk, getSignature());
}

bool GetAccounts::checkSignature(const uint8_t* /*hash*/, const uint8_t* pk) {
    return (ed25519_sign_open(getData(), getDataSize(), pk, getSignature()) == 0);
}

void GetAccounts::saveResponse(settings& /*sts*/) {
}

uint32_t GetAccounts::getUserId() {
    return m_data.info.src_user;
}

uint32_t GetAccounts::getBankId() {
    return m_data.info.src_node;
}

uint32_t GetAccounts::getTime() {
    return m_data.info.ttime;
}

int64_t GetAccounts::getFee() {
    return 0;
}

int64_t GetAccounts::getDeduct() {
    return 0;
}

bool GetAccounts::send(INetworkClient& netClient) {
    if(!netClient.sendData(getData(), sizeof(m_data))) {
        ELOG("GetAccounts sending error\n");
        return false;
    }

    if (!netClient.readData((int32_t*)&m_responseError, ERROR_CODE_LENGTH)) {
        ELOG("GetAccounts reading error\n");
        return false;
    }

    if (m_responseError) {
        return true;
    }

    if (!netClient.readData((int32_t*)&m_responseBufferLength, sizeof(uint32_t))) {
        ELOG("Get accounts reading error\n");
        return false;
    }

    // read rest of buffer, all accounts
    m_responseBuffer = std::make_unique<unsigned char[] >(m_responseBufferLength);
    if (!netClient.readData(m_responseBuffer.get(), m_responseBufferLength)) {
        ELOG("GetAccounts ERROR reading response\n");
        return false;
    }
    return true;
}

uint32_t GetAccounts::getDestBankId() {
    return m_data.info.dst_node;
}

uint32_t GetAccounts::getBlockId() {
    return m_data.info.block;
}

uint32_t GetAccounts::getUserMessageId() {
    return 0;
}

std::string GetAccounts::toString(bool /*pretty*/) {
    return "";
}

void GetAccounts::toJson(boost::property_tree::ptree& ptree) {
    if (!m_responseError) {
        uint32_t no_of_users = this->getResponseSize() / sizeof(user_t);
        user_t* user_ptr=(user_t*)this->getResponse();
        uint32_t bankId = this->getDestBankId();
        boost::property_tree::ptree users;
        for(uint32_t i=0; i<no_of_users; i++, user_ptr++) {
            boost::property_tree::ptree user;
            print_user(*user_ptr, user, true, bankId, i);
            users.push_back(std::make_pair("", user.get_child("account")));
        }
        ptree.put_child("accounts", users);
    } else {
        ptree.put(ERROR_TAG, ErrorCodes().getErrorMsg(m_responseError));
    }
}

ErrorCodes::Code GetAccounts::prepareResponse(uint32_t lastPath, uint32_t lastUsers) {
    ErrorCodes::Code errorCode = ErrorCodes::Code::eNone;
    uint32_t destBank = this->getDestBankId();
    char filename[64];
    int fd;
    std::shared_ptr<Helper::BlockFileReader> undo;

    sprintf(filename,"usr/%04X.dat", destBank);
    fd = open(filename,O_RDONLY);
    if (fd < 0) {
        errorCode = ErrorCodes::Code::eBankNotFound;
    } else {
        Helper::FileName::getUndo(filename, lastPath, destBank);
        undo = std::make_shared<Helper::BlockFileReader>(filename);
        if(!undo->isOpen()) {
            errorCode = ErrorCodes::Code::eUndoNotFound;
        }
    }

    if (!errorCode) {
        m_responseBufferLength = lastUsers * sizeof(user_t);
        //m_responseBuffer = new unsigned char[m_responseBufferLength];
        m_responseBuffer = std::make_unique<unsigned char[] >(m_responseBufferLength);
        uint8_t *dp = m_responseBuffer.get();
        for(uint32_t user=0; user<lastUsers; user++,dp+=sizeof(user_t)) {
            user_t u;
            u.msid=0;
            read(fd,dp,sizeof(user_t));
            undo->read(&u,sizeof(user_t));
            if(u.msid!=0) {
                memcpy(dp,&u,sizeof(user_t));
            }
        }
    }
    close(fd);

    return errorCode;
}

void GetAccounts::txnToJson(boost::property_tree::ptree& ptree) {
    using namespace Helper;
    ptree.put(TAG::TYPE, getTxnName(m_data.info.ttype));
    ptree.put(TAG::SRC_NODE, m_data.info.src_node);
    ptree.put(TAG::SRC_USER, m_data.info.src_user);
    ptree.put(TAG::DST_NODE, m_data.info.dst_node);
    ptree.put(TAG::TIME, m_data.info.ttime);
    ptree.put(TAG::BLOCK, m_data.info.block);
    ptree.put(TAG::SIGN, ed25519_key2text(getSignature(), getSignatureSize()));
}

std::string GetAccounts::usageHelperToString() {
    std::stringstream ss{};
    ss << "Usage: " << "{\"run\":\"get_accounts\",[\"node\":<node id>],[\"block\":<block_time as hex>]}" << "\n";
    ss << "Example: " << "{\"run\":\"get_accounts\",\"node\":\"1\",\"block\":\"5B052CE0\"}" << "\n";
    return ss.str();
}
